File: dialogs/src/main/java/com/rands/lightbulb/dialogs/base/constraints/regular/RegularDialogConstraintsBuilder.java
package com.rands.lightbulb.dialogs.base.constraints.regular;

import android.content.res.Configuration;

import com.rands.lightbulb.dialogs.base.BaseDialogFragment;
import com.rands.lightbulb.dialogs.base.constraints.base.BaseDialogConstraintBuilder;

@SuppressWarnings("FieldCanBeLocal")
public class RegularDialogConstraintsBuilder extends BaseDialogConstraintBuilder {
    private final int DEF_MAX_WIDTH_PERCENTS_PORTRAIT = 85;
    private final int DEF_MAX_WIDTH_PERCENTS_LANDSCAPE = 65;
    private final int DEF_MAX_HEIGHT_PERCENTS = 85;
    private int minWidth;
    private int maxWidth;
    private int minHeight;
    private int maxHeight;

    public RegularDialogConstraintsBuilder(BaseDialogFragment dialogFragment) {
        super(dialogFragment);
    }

    public RegularDialogConstraintsBuilder Default() {
        int orientation = dialogFragment.getResources().getConfiguration().orientation;
        int maxWidth = getPercentOfWindowWidth(
                orientation == Configuration.ORIENTATION_PORTRAIT ?
                        DEF_MAX_WIDTH_PERCENTS_PORTRAIT : DEF_MAX_WIDTH_PERCENTS_LANDSCAPE
        );
        int maxHeight = getPercentOfWindowHeight(DEF_MAX_HEIGHT_PERCENTS);
        this.minWidth = 0;
        this.maxWidth = maxWidth;
        this.minHeight = 0;
        this.maxHeight = maxHeight;
        return this;
    }

    public RegularDialogConstraintsBuilder withMinWidth(int minWidth) {
        this.minWidth = minWidth;
        return this;
    }

    public RegularDialogConstraintsBuilder withMaxWidth(int maxWidth) {
        this.maxWidth = maxWidth;
        return this;
    }

    public RegularDialogConstraintsBuilder withMinHeight(int minHeight) {
        this.minHeight = minHeight;
        return this;
    }

    public RegularDialogConstraintsBuilder withMaxHeight(int maxHeight) {
        this.maxHeight = maxHeight;
        return this;
    }

    public RegularDialogConstraintsBuilder withMinWidthInPercents(int minWidthPercent) {
        minWidthPercent = validatePercentNumber(minWidthPercent);
        minWidth = getPercentOfWindowWidth(minWidthPercent);
        return this;
    }

    public RegularDialogConstraintsBuilder withMaxWidthInPercents(int maxWidthPercent) {
        maxWidthPercent = validatePercentNumber(maxWidthPercent);
        maxWidth = getPercentOfWindowWidth(maxWidthPercent);
        return this;
    }

    public RegularDialogConstraintsBuilder withMinHeightInPercents(int minHeightPercent) {
        minHeightPercent = validatePercentNumber(minHeightPercent);
        minHeight = getPercentOfWindowHeight(minHeightPercent);
        return this;
    }

    public RegularDialogConstraintsBuilder withMaxHeightInPercents(int maxHeightPercent) {
        maxHeightPercent = validatePercentNumber(maxHeightPercent);
        maxHeight = getPercentOfWindowHeight(maxHeightPercent);
        return this;
    }

    public RegularDialogConstraints build() {
        return new RegularDialogConstraints(minWidth, maxWidth, minHeight, maxHeight);
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadowss/lightbulb/dialogs/base/constraints/regular/RegularDialogConstraintsBuilder.kt
package com.github.rooneyandshadowss.lightbulb.dialogs.base.constraints.regular

import android.content.res.Configuration
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment
import com.github.rooneyandshadows.lightbulb.dialogs.base.constraints.base.BaseDialogConstraintBuilder

class RegularDialogConstraintsBuilder(dialogFragment: BaseDialogFragment) : BaseDialogConstraintBuilder(dialogFragment) {
    private val defaultMaxWidthPercentsPortrait = 85
    private val defaultMaxWidthPercentsLandscape = 65
    private val defaultMaxHeightPercents = 85
    private var minWidth = 0
    private var maxWidth = 0
    private var minHeight = 0
    private var maxHeight = 0

    fun default(): RegularDialogConstraintsBuilder {
        val orientation = dialogFragment.resources.configuration.orientation
        val maxWidth = getPercentOfWindowWidth(
            if (orientation == Configuration.ORIENTATION_PORTRAIT)
                defaultMaxWidthPercentsPortrait
            else
                defaultMaxWidthPercentsLandscape
        )
        val maxHeight = getPercentOfWindowHeight(defaultMaxHeightPercents)
        minWidth = 0
        this.maxWidth = maxWidth
        minHeight = 0
        this.maxHeight = maxHeight
        return this
    }

    fun withMinWidth(minWidth: Int): RegularDialogConstraintsBuilder {
        this.minWidth = minWidth
        return this
    }

    fun withMaxWidth(maxWidth: Int): RegularDialogConstraintsBuilder {
        this.maxWidth = maxWidth
        return this
    }

    fun withMinHeight(minHeight: Int): RegularDialogConstraintsBuilder {
        this.minHeight = minHeight
        return this
    }

    fun withMaxHeight(maxHeight: Int): RegularDialogConstraintsBuilder {
        this.maxHeight = maxHeight
        return this
    }

    fun withMinWidthInPercents(minWidthPercent: Int): RegularDialogConstraintsBuilder {
        minWidth = getPercentOfWindowWidth(validatePercentNumber(minWidthPercent))
        return this
    }

    fun withMaxWidthInPercents(maxWidthPercent: Int): RegularDialogConstraintsBuilder {
        maxWidth = getPercentOfWindowWidth(validatePercentNumber(maxWidthPercent))
        return this
    }

    fun withMinHeightInPercents(minHeightPercent: Int): RegularDialogConstraintsBuilder {
        minHeight = getPercentOfWindowHeight(validatePercentNumber(minHeightPercent))
        return this
    }

    fun withMaxHeightInPercents(maxHeightPercent: Int): RegularDialogConstraintsBuilder {
        maxHeight = getPercentOfWindowHeight(validatePercentNumber(maxHeightPercent))
        return this
    }

    fun build(): RegularDialogConstraints {
        return RegularDialogConstraints(minWidth, maxWidth, minHeight, maxHeight)
    }
}